### PR TITLE
Adding trimming compatibility

### PR DESCRIPTION
**Pull request by mbannick**
- Added a trimming compatible model class. Main thing is that objective function and gradient need to be decomposable by observation
- Added a composite solver for trimming. Utilizes the trimming compatible model class.

### CODE REVIEW

**Comment by mbannick** [comment]
Ignore "hack" commits. I'll remove those later before merging. Just the trimming stuff now.

**Comment by mbannick** [comment]
@zhengp0 let me know if you think I can merge!

File: src/anml/models/interface.py [hunk 1]
```diff
@@ -78,3 +79,61 @@ def forward(self, x: np.ndarray, *args, **kwargs):
         # different from predict() in solver in the sense that both variable and data value can vary.
         # in predict() variable value is at taken to be the optimal.
         raise NotImplementedError()
+
+
+class TrimmingCompatibleModel(Model, ABC):
+
+    def __init__(self):
+        super().__init__()
```
Comment (zhengp0) [inline src/anml/models/interface.py:87]: Maybe we could move `self._w` from concrete class to this abstract class :)

File: src/anml/solvers/base.py [hunk 2]
```diff
@@ -6,7 +6,7 @@
 Basic solvers for optimization. Takes a model directly. 
 """
 
-import ipopt
+# import ipopt
```
Comment (zhengp0) [inline src/anml/solvers/base.py:9]: Thank you! I think we need to get rid of this at some point :)

File: src/anml/solvers/composite.py [hunk 3]
```diff
@@ -38,3 +40,67 @@ def fit(self, x_init: Optional[np.ndarray] = None, data: Optional[Data] = None,
...
+
+        i = 0
+
+        if pct_trimming > 0:
+
+            # Initialize the weights and trimming model
+            w_init = np.repeat(pct_trimming, n)
+            h = (1 - pct_trimming) * n
+
+            iter_count = 0
```
Comment (zhengp0) [inline src/anml/solvers/composite.py:82]: `iter_count` and `i` are the same, we could just keep one of them.

**Review by zhengp0 — APPROVED** [review]
Thanks @mbannick ! Some minor comments, and I think it is good!

File: src/anml/models/interface.py (hunk 1 above)
Comment (mbannick) [reply src/anml/models/interface.py:87]: Yes, good idea!

**Pull request merged**